### PR TITLE
refactor: Kotlin migration of CardBrowser 

### PR DESCRIPTION
**Pull request by david-allison**
Briefly tested on my Android. Made corrections to fix tests

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by Akshay0701** [comment]
Wow!!!!!

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 1]
```diff
@@ -0,0 +1,2744 @@
...
+                if (card.id == mOldCardId) {
+                    return card.position
+                }
+            }
+            return CARD_NOT_AVAILABLE
+        }
+
+    fun hasSelectedAllDecks(): Boolean {
+        val lastDeckId = lastDeckId
+        return lastDeckId != null && lastDeckId == ALL_DECKS_ID
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:1903]: why do we create a separate variable just to check the below conditions?

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 2]
```diff
@@ -0,0 +1,2744 @@
...
+        override fun getItem(position: Int): CardCache {
+            return cards[position]
+        }
+
+        override fun getItemId(position: Int): Long {
+            return position.toLong()
+        }
+
+        init {
+            if ("" != customFont) {
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:2178]: `isNotEmpty` would be good

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 3]
```diff
@@ -0,0 +1,2744 @@
...
+    @VisibleForTesting
+    fun onSelectNone() {
+        mCheckedCards.clear()
+        onSelectionChanged()
+    }
+
+    private fun onSelectionChanged() {
+        Timber.d("onSelectionChanged()")
+        try {
+            if (!isInMultiSelectMode && !mCheckedCards.isEmpty()) {
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:2211]: isNotEmpty

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 4]
```diff
@@ -0,0 +1,2744 @@
...
+    @VisibleForTesting
+    fun cardCount(): Long {
+        return mCards.size().toLong()
+    }
+
+    @get:VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    val isShowingSelectAll: Boolean
+        get() = mActionBarMenu != null && mActionBarMenu!!.findItem(R.id.action_select_all).isVisible
+
+    //
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:2560]: revert this change

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 5]
```diff
@@ -0,0 +1,2744 @@
...
+            for (card in mCheckedCards) {
+                val id = card.id
+                cardIds.add(id)
+            }
+            return cardIds
+        }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    fun // should only be called from changeDeck()
+    executeChangeCollectionTask(ids: List<Long>?, newDid: Long) {
```
Comment (Akshay0701) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:2619]:
```suggestion
    // should only be called from changeDeck()
    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
    fun executeChangeCollectionTask(ids: List<Long>?, newDid: Long) {
```
maybe?

**Review by Akshay0701 — CHANGES_REQUESTED** [review]
Great work

**Comment by Akshay0701** [comment]
Sure, will do from next time, Thanks!

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 3 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:2211]: handled by IDE lint

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 4 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:2560]: Do you mean the extra `//`? If so, then done

**Comment by david-allison** [comment]
@Akshay0701 one query, done the rest, thanks for picking up on the laziness :)

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 4 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt]: my bad, yes i meaned the comment //

**Comment by david-allison** [comment]
Akshay's answered, all comments dealt with

**Review by mikehardy — APPROVED** [review]
The class itself is ridiculous, but the migration (with associated cleanups as noted) LGTM

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 6]
```diff
@@ -0,0 +1,2747 @@
...
+        if (deck == null) {
+            return
+        }
+        val deckId = deck.deckId
+        mDeckSpinnerSelection!!.initializeActionBarDeckSpinner(this.supportActionBar!!)
+        mDeckSpinnerSelection!!.selectDeckById(deckId, true)
+        selectDeckAndSave(deckId)
+    }
+
+    enum class Column {
```
Comment (raymondwzeng) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:122]: Minor nitpick: does this exceed the 120-column guideline from the code style?

**Review by raymondwzeng — APPROVED** [review]
Looks good to me! Pleasantly surprised that my work ended up very similar to the accepted PR, with most of the differences coming from the transformation of Collection types (IDE might have gone a _bit_ wild on the nullable function signature on my end) and null checks (throwing NPE with !! vs ignoring with ?, and it seems like this PR handles nullable args much more elegantly). Also good to learn about handling annotations at the top of code blocks, for future reference.

Changing the DeckSelectionDialog to a functional interface was a good call. I was mulling over it, but ultimately ended up not pulling the trigger due to not quite knowing whether or not it would affect the Java classes that still used it. Guess I underestimated the interoperability of Kotlin and Java.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 6 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:122]: At this point - with kotlin - thankfully it's not just a guide, it's enforced by ktlint. We never got around to integrating google-java-format or any auto-linters/formatters for java, but for kotlin we're starting it out right.

Which is to say: it's either enforced by lint or not. I personally don't see a big problem with long line lengths anymore within reason - that is, they improve readability. Certainly 80chars is dead and what is 120 anyway? So I'm not bothered enough to change ktlint definition to enforce line length.

But these things are a matter of taste (I think not technical anymore like 80char terminals...) and it's just a personal preference. Others can disagree and if there's consensus we can add a line length to ktlint.

Comment (raymondwzeng) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:122]: Oh wow, I didn't know that there was even a technical explanation/origin for column limits per line.

That aside, I agree with your point that it depends on readability and varies from person to person. And as much as I do like things to fit without scrolling horizontally (which upon further evaluation, this line does), that would be a nontrivial thing to determine a standard for given aspect ratios and physical display sizes.
